docs/extending_fuel.rst

```diff
@@ -0,0 +1,268 @@
+with ``.npy`` files.
+
+>>> from collections import OrderedDict
+>>> import numpy
+>>> from six import iteritems
+>>> from fuel.datasets import IndexableDataset
+>>>
+>>> class NPYDataset(IndexableDataset):
+...     def __init__(self, source_paths, **kwargs):
+...         indexables = OrderedDict(
+...             [(source, numpy.load(path)) for
+...              source, path in iteritems(source_paths)])
+...         super(NPYDataset, self).__init__(indexables, **kwargs)
+
+Here's this class in action:
+
+>>> numpy.save('npy_dataset_features.npy',
+...            numpy.arange(40).reshape((10, 4)))
+>>> numpy.save('npy_dataset_targets.npy',
+...            numpy.arange(10).reshape((10, 1)))
+>>> dataset = NPYDataset(OrderedDict([('features', 'npy_dataset_features.npy'),
+...                                   ('targets', 'npy_dataset_targets.npy')]))
+>>> state = dataset.open()
+>>> print(dataset.get_data(state=state, request=[0, 1, 2, 3]))
+(array([[ 0,  1,  2,  3],
+       [ 4,  5,  6,  7],
+       [ 8,  9, 10, 11],
+       [12, 13, 14, 15]]), array([[0],
+       [1],
+       [2],
+       [3]]))
+>>> dataset.close(state)
+
+.. doctest::
+   :hide:
+
+   >>> import os
+   >>> os.remove('npy_dataset_features.npy')
+   >>> os.remove('npy_dataset_targets.npy')
+
+New transformers
+----------------
+
+An important thing to know about data streams is that they distinguish
+between two types of outputs: single examples, and batches of examples.
+Depending on your choice of iteration scheme, a data stream's
+``produces_examples`` property will either be ``True`` (it produces examples)
+or ``False`` (it produces batches).
+
+Transformers are aware of this, and as such implement two distinct methods:
+:meth:`transform_example` and :meth:`transform_batch`. A new transformer is
+typically implemented by subclassing Transformer and implementing one or both
+of these methods.
+
+As an example, here's how you would double the value of some ``'features'``
+data source.
+
+>>> from fuel.transformers import Transformer
+>>>
+>>> class FeaturesDoubler(Transformer):
+...     def __init__(self, data_stream, **kwargs):
+...         super(FeaturesDoubler, self).__init__(
+...             data_stream=data_stream,
+...             produces_examples=data_stream.produces_examples,
+...             **kwargs)
+...         
+...     def transform_example(self, example):
+...         if 'features' in self.sources:
+...             example = list(example)
+...             index = self.sources.index('features')
+...             example[index] *= 2
+...             example = tuple(example)
+...         return example
+...     
+...     def transform_batch(self, batch):
+...         if 'features' in self.sources:
+...             batch = list(batch)
+...             index = self.sources.index('features')
+...             batch[index] *= 2
+...             batch = tuple(batch)
+...         return batch
+
+Most transformers you'll implement will call their superclass constructor by
```

Comment:
This paragraph is a non-mandatory best practice. It could be separated from the rest of the text and added as a "Hint box" at the end of this section.